### PR TITLE
Sort during find results

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo/issues/946

### CODE REVIEW

File: neo/IO/Caching/DataCache.cs [hunk 1]
```diff
@@ -119,6 +120,47 @@ public void DeleteWhere(Func<TKey, TValue, bool> predicate)
         }
 
         public IEnumerable<KeyValuePair<TKey, TValue>> Find(byte[] key_prefix = null)
+        {
+            lock (dictionary)
+            {
+                return FindUnsorted(key_prefix).OrderBy(u => u.Key, this);
```
Comment (erikzhang) [inline neo/IO/Caching/DataCache.cs]: This forces the entire set of search results to be loaded into memory for sorting, which can lead to attacks.

Comment (vncoelho) [reply neo/IO/Caching/DataCache.cs]: Still nowadays I have dificulties with C#.

Looking at it I would think that find it finds and returns the set, then, it sorts the results.

In this sense, maybe it needs to be split into two steps, right, @erikzhang?

Comment (erikzhang) [reply neo/IO/Caching/DataCache.cs]: We need to do some optimization. Because the records read from the underlying storage (leveldb) is already sorted.

Comment (vncoelho) [reply neo/IO/Caching/DataCache.cs]: Perfect, master.
We are going to reflect on that we think about an efficient way.

Comment (shargon) [reply neo/IO/Caching/DataCache.cs]: I will propose another way

File: neo/IO/Caching/DataCache.cs [hunk 2]
```diff
@@ -122,13 +123,122 @@ public void DeleteWhere(Func<TKey, TValue, bool> predicate)
...
+
+                        if (iterator.MoveNext())
+                        {
+                            a = iterator.Current;
+                            if (iterator.MoveNext()) b = iterator.Current;
+                        }
+
+                        // We need to return a, b and cached results between them
+
+                        if (a.Key == null && a.Value == null)
```
Comment (shargon) [inline neo/IO/Caching/DataCache.cs]: ReferenceEquals(a,n) doesn't work :S

**Comment by erikzhang** [comment]
@shargon I implemented a more optimized fix, please take a look. #950

**Pull request closed without merging**